File: crates/prequel-capture/src/wallpaper.rs
//! Capturing the desktop picture as it currently looks.
//!
//! Reading the wallpaper *file* on macOS 14+ is unreliable: the legacy
//! `desktoppicture.db` is often gone, the current store leaves its file list
//! empty for the stock pictures and hides the identity in an NSKeyedArchiver
//! blob, `NSWorkspace.desktopImageURL(for:)` is unbound and returns a
//! multi-image dynamic HEIC anyway, and `osascript` needs an Automation grant
//! that fails confusingly.
//!
//! Screenshotting the wallpaper window sidesteps all of it, and Prequel is
//! unusually well placed to do so: it already holds the Screen Recording grant
//! and cannot function without one. It is also the only approach that is
//! correct for dynamic and video wallpapers, where there is no still file to
//! find — what you get is what is on screen.

use std::path::Path;
use std::sync::mpsc;
use std::time::Duration;

use cidre::{cg, ci, cm, ns, sc};

use crate::{Error, Result};

/// The agent that draws the desktop picture.
const WALLPAPER_BUNDLE_ID: &str = "com.apple.wallpaper.agent";

/// Also matched on layer, because the agent's bundle id has changed between
/// macOS releases and the wallpaper is always at the very back.
const DESKTOP_LAYER: ns::Integer = -2_147_483_648;

const CAPTURE_TIMEOUT: Duration = Duration::from_secs(10);

/// Writes the current desktop picture of `display_id` to `path`, as a PNG.
///
/// `display_id` is a `CGDirectDisplayID`; zero means the main display.
pub fn capture_wallpaper(display_id: u32, path: &Path) -> Result<()> {
    let content = crate::targets::current_shareable_content()?;

    let displays = content.displays();
    let display = displays
        .iter()
        .find(|candidate| {
            display_id == 0 || candidate.display_id() == cg::DirectDisplayId(display_id)
        })
        .or_else(|| displays.iter().next())
        .ok_or(Error::DisplayNotFound(display_id))?;

    let windows = content.windows();
    let wallpaper = windows
        .iter()
        .find(|window| {
            window
                .owning_app()
                .is_some_and(|app| app.bundle_id().to_string() == WALLPAPER_BUNDLE_ID)
        })
        // The desktop is the bottom-most layer, which is what identifies it
        // when the bundle id does not.
        .or_else(|| {
            windows
                .iter()
                .filter(|window| window.window_layer() <= DESKTOP_LAYER)
                .min_by_key(|window| window.window_layer())
        })
        .ok_or_else(|| Error::ScreenCaptureKit("no wallpaper window on screen".to_owned()))?;

    let filter = sc::ContentFilter::with_desktop_independent_window(wallpaper);

    let mut cfg = sc::StreamCfg::new();
    cfg.set_width(display.width() as usize);
    cfg.set_height(display.height() as usize);

    let sample = capture(&filter, &cfg)?;
    write_png(&sample, path)
}

/// Takes one frame, blocking until ScreenCaptureKit hands it over.
///
/// Block-based rather than async: this is called from a worker thread whose
/// whole job is to produce the file, and introducing a runtime to await one
/// callback would be a lot of machinery for no benefit.
fn capture(
    filter: &sc::ContentFilter,
    cfg: &sc::StreamCfg,
) -> Result<cidre::arc::R<cm::SampleBuf>> {
    let (tx, rx) = mpsc::channel();

    let mut handler = cidre::blocks::ResultCh::new2(
        move |sample: Option<&cm::SampleBuf>, error: Option<&ns::Error>| {
            let _ = tx.send(match (sample, error) {
                (Some(sample), _) => Ok(sample.retained()),
                (None, Some(error)) => Err(error.to_string()),
                (None, None) => Err("no image and no error".to_owned()),
            });
        },
    );

    sc::ScreenshotManager::capture_sample_buf_ch(filter, cfg, Some(&mut handler));

    rx.recv_timeout(CAPTURE_TIMEOUT)
        .map_err(|_| Error::Timeout(CAPTURE_TIMEOUT))?
        .map_err(Error::ScreenCaptureKit)
}

/// Writes a captured frame out as a PNG.
///
/// Through Core Image because it is the one path cidre binds end to end —
/// `ci::Context::write_png_to_url` — and this runs once per background choice,
/// so nothing here is on a hot path.
fn write_png(sample: &cm::SampleBuf, path: &Path) -> Result<()> {
    let buffer = sample
        .image_buf()
        .ok_or_else(|| Error::ScreenCaptureKit("the captured frame carried no image".to_owned()))?;

    let image = ci::Image::with_cv_image_buf(buffer, None)
        .ok_or_else(|| Error::ScreenCaptureKit("could not read the captured frame".to_owned()))?;

    let path = path
        .to_str()
        .ok_or_else(|| Error::ScreenCaptureKit("output path is not valid UTF-8".to_owned()))?;

    let context = ci::Context::new();
    let url = ns::Url::with_fs_path_str(path, false);
    let color_space = cg::ColorSpace::device_rgb()
        .ok_or_else(|| Error::ScreenCaptureKit("no device colour space".to_owned()))?;

    context
        .write_png_to_url(
            &image,
            &url,
            ci::Format::rgba8(),
            &color_space,
            &ns::Dictionary::new(),
        )
        .map_err(|err| Error::ScreenCaptureKit(err.to_string()))
}

File: apps/desktop/src/main/wallpaper.ts
/**
 * Backgrounds, copied into the recording that uses them.
 *
 * Copied rather than referenced, and that is the important part: the desktop
 * picture changes tomorrow, and an export has to produce the same video it
 * produced today. It also makes a recording directory self-contained, which is
 * what lets it move between machines.
 *
 * Reading the *current* wallpaper is harder than it should be on macOS 14+.
 * `~/Library/Application Support/Dock/desktoppicture.db` is legacy and often
 * absent; the current store, `com.apple.wallpaper/Store/Index.plist`, leaves
 * its `Files` array empty for the stock pictures and hides the real identity
 * inside an NSKeyedArchiver blob; `NSWorkspace.desktopImageURL(for:)` is not
 * bound by cidre and returns a multi-image dynamic HEIC anyway; and
 * `osascript` needs an Automation grant that fails confusingly.
 *
 * So the plist is only a fast path, for the case where the user picked their
 * own image and it is named outright. Anything else falls back to a real
 * screenshot of the wallpaper window, which Prequel is uniquely able to take
 * because it already holds the Screen Recording grant — and which is correct
 * for dynamic and video wallpapers, where there is no still file to find.
 */
import { execFile } from "node:child_process";
import { copyFileSync, existsSync } from "node:fs";
import { homedir } from "node:os";
import { join } from "node:path";
import { fileURLToPath } from "node:url";
import { promisify } from "node:util";

import { dialog } from "electron";

import { backgroundPreset } from "../shared/backgrounds.js";
import type { BackgroundImage } from "../shared/contract.js";
import { mediaUrl } from "./media-protocol.js";
import { getRecorder } from "./recorder.js";

const run = promisify(execFile);

import { WALLPAPER_FILE_NAME as WALLPAPER_FILE } from "../shared/project.js";

/** Longest edge of a copied background. A 6K wallpaper is not worth keeping. */
const MAX_EDGE = 2560;

const WALLPAPER_INDEX = join(
  homedir(),
  "Library",
  "Application Support",
  "com.apple.wallpaper",
  "Store",
  "Index.plist",
);

/**
 * Puts the current desktop picture into a recording.
 *
 * Returns null when it cannot be found at all, so the option can simply be
 * absent rather than failing — the same posture `app-icons.ts` takes, where a
 * sandbox refusal is not worth failing the whole picker over.
 */
export async function captureWallpaper(dir: string): Promise<BackgroundImage | null> {
  return capture(dir, { reuse: false });
}

/**
 * Makes sure a recording has a wallpaper to use as its default background.
 *
 * Reuses the copy already in the directory rather than re-capturing: the
 * default is meant to be what the desktop looked like when the recording was
 * made, and re-taking it on every open would silently restyle an old edit the
 * next time the user changed their wallpaper.
 */
export async function ensureWallpaper(dir: string): Promise<boolean> {
  return (await capture(dir, { reuse: true })) !== null;
}

async function capture(dir: string, options: { reuse: boolean }): Promise<BackgroundImage | null> {
  const destination = join(dir, WALLPAPER_FILE);

  if (options.reuse && existsSync(destination)) return described(dir, WALLPAPER_FILE);

  const named = await namedWallpaperFile();
  if (named) {
    // `sips` rather than a straight copy: the stock pictures are HEIC, which a
    // renderer cannot draw, and they are far larger than a background needs.
    if (await convert(named, destination)) return described(dir, WALLPAPER_FILE);
  }

  // No file to find — a stock picture, or a dynamic or video wallpaper that has
  // no still image at all. Screenshot what is actually on the desktop instead.
  try {
    await (await getRecorder()).captureWallpaper(0, destination);
    // Downscaled in place: the capture comes back at the display's full
    // resolution, which on a Retina panel is far more than a background needs.
    await convert(destination, destination);
    return described(dir, WALLPAPER_FILE);
  } catch (cause) {
    // Not worth failing the picker over — the option is simply absent and the
    // other backgrounds still work.
    console.warn("[wallpaper] could not capture the desktop:", cause);
    return null;
  }
}

/**
 * Lets the user choose their own background image.
 *
 * Copied in under a stable name so the project can reference it relatively.
 */
export async function pickBackgroundImage(dir: string): Promise<BackgroundImage | null> {
  const { canceled, filePaths } = await dialog.showOpenDialog({
    title: "Choose a background",
    properties: ["openFile"],
    filters: [{ name: "Images", extensions: ["png", "jpg", "jpeg", "heic", "heif"] }],
  });

  const source = filePaths[0];
  if (canceled || !source) return null;

  const name = "background-custom.png";
  const destination = join(dir, name);

  // Converted rather than copied: a HEIC or a 48-megapixel photo would either
  // fail to draw or make every preview frame expensive.
  if (await convert(source, destination)) return described(dir, name);

  try {
    copyFileSync(source, destination);
    return described(dir, name);
  } catch (cause) {
    console.warn("[wallpaper] could not copy the chosen image:", cause);
    return null;
  }
}

/**
 * Copies one of the shipped wallpapers into the recording.
 *
 * Copied rather than referenced, for the same reason the desktop picture is:
 * a recording directory that carries everything it needs can be moved between
 * machines and will still export the video it exported today. The app's own
 * asset could also change between releases, and an edit should not.
 *
 * Already downscaled and compressed on the way into the bundle, so unlike a
 * user-chosen photo there is nothing here worth converting.
 */
export function copyPresetBackground(dir: string, presetId: string): BackgroundImage | null {
  const preset = backgroundPreset(presetId);
  if (!preset) {
    console.warn(`[wallpaper] no such background preset: ${presetId}`);
    return null;
  }

  const destination = join(dir, preset.file);

  try {
    // Skipped when it is already there — picking the same background twice is
    // not a reason to rewrite a megabyte.
    if (!existsSync(destination)) {
      copyFileSync(
        fileURLToPath(new URL(`../../resources/backgrounds/${preset.file}`, import.meta.url)),
        destination,
      );
    }
    return described(dir, preset.file);
  } catch (cause) {
    console.warn(`[wallpaper] could not copy ${preset.file}:`, cause);
    return null;
  }
}

/**
 * The wallpaper's own file, when the store names one.
 *
 * Only true when the user picked their own picture. The stock ones report a
 * `default` provider with an empty file list, which is why this is a fast path
 * rather than the mechanism.
 */
async function namedWallpaperFile(): Promise<string | null> {
  if (!existsSync(WALLPAPER_INDEX)) return null;

  try {
    // The same tool `app-icons.ts` uses, for the same reason: these are binary
    // plists, and nothing will read them as text.
    //
    // XML rather than JSON. The store keeps an NSKeyedArchiver blob beside the
    // file list, JSON has no way to spell a `data` value, and `plutil` refuses
    // the *whole* conversion over it: "Invalid object in plist for JSON
    // format". So the fast path did not merely miss the odd wallpaper, it threw
    // on every Mac whose store holds a blob — which is every Mac using a stock
    // picture — and the warning it logged said nothing about why.
    const { stdout } = await run("/usr/bin/plutil", [
      "-convert",
      "xml1",
      "-o",
      "-",
      WALLPAPER_INDEX,
    ]);
    const found = [...stdout.matchAll(/<string>(file:[^<]+)<\/string>/g)]
      .map((match) => match[1]!)
      .map((url) => {
        try {
          return decodeURIComponent(new URL(url).pathname);
        } catch {
          return null;
        }
      })
      .find((path): path is string => path !== null && existsSync(path));

    return found ?? null;
  } catch (cause) {
    console.warn("[wallpaper] could not read the wallpaper store:", cause);
    return null;
  }
}

/** Flattens and downscales to a PNG the renderer can draw cheaply. */
async function convert(source: string, destination: string): Promise<boolean> {
  try {
    await run("/usr/bin/sips", [
      "-s",
      "format",
      "png",
      "-Z",
      String(MAX_EDGE),
      source,
      "--out",
      destination,
    ]);
    return existsSync(destination);
  } catch (cause) {
    console.warn(`[wallpaper] could not convert ${source}:`, cause);
    return false;
  }
}

function described(dir: string, name: string): BackgroundImage {
  return { path: name, url: mediaUrl(dir, name) };
}
